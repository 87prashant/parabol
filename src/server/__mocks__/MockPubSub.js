import * as getPubSub from 'server/utils/getPubSub';

const fieldsToSerialize = {
  invitation: [
    'channelId',
    'message.data.invitation.id',
    'message.data.invitation.email',
    'message.data.invitation.hashedToken',
    'message.data.invitation.id',
    'message.data.invitation.invitedBy',
    'message.data.invitation.teamId'
  ],
  invitationAdded: [
    'channelId',
    'message.invitationAdded.invitation.id',
    'message.invitationAdded.invitation.email',
    'message.invitationAdded.invitation.hashedToken',
    'message.invitationAdded.invitation.id',
    'message.invitationAdded.invitation.invitedBy',
    'message.invitationAdded.invitation.teamId'
  ],
  invitationRemoved: [
    'channelId',
    'message.invitationRemoved.invitation.id',
    'message.invitationRemoved.invitation.email',
    'message.invitationRemoved.invitation.hashedToken',
    'message.invitationRemoved.invitation.inviteToken',
    'message.invitationRemoved.invitation.invitedBy',
    'message.invitationRemoved.invitation.teamId'
  ],
  newAuthToken: [
    'channelId',
    'message.newAuthToken'
  ],
  notification: [
    'channelId',
    'message.data.notifications.id',
    'message.data.notifications.orgId',
    'message.data.notifications.teamId',
    'message.data.notifications.userIds',
    'message.data.notifications.inviteeEmail'
  ],
  notificationsCleared: [
    'channelId',
    'message.notificationsCleared.deletedIds'
  ],
  orgApprovalRemoved: [
    'channelId',
    'message.orgApprovalRemoved.orgApproval.approvedBy',
    'message.orgApprovalRemoved.orgApproval.deniedBy',
    'message.orgApprovalRemoved.orgApproval.email',
    'message.orgApprovalRemoved.orgApproval.id',
    'message.orgApprovalRemoved.orgApproval.orgId',
    'message.orgApprovalRemoved.orgApproval.teamId'
  ],
  teamAdded: [
    'channelId',
    'message.teamAdded.notification.teamId',
    'message.teamAdded.teamId',
    'message.teamAdded.teamMemberId'
  ],
  teamUpdated: [
    'channelId',
    'message.teamUpdated.team.id',
    'message.teamUpdated.team.orgId'
  ],
  teamMemberAdded: [
    'channelId',
    'message.teamMemberAdded.teamMember.id',
    'message.teamMemberAdded.teamMember.email',
    'message.teamMemberAdded.teamMember.teamId',
    'message.teamMemberAdded.teamMember.userId'
  ]
};

const fieldsToSort = {
  notificationsAdded: [
    'message.notificationsAdded.notifications.0.teamName',
    'message.notificationsAdded.notifications.0.type'
  ],
  orgApprovalRemoved: [
    'message.orgApprovalRemoved.orgApproval.createdAt'
  ]
};

const unsortables = new Set(['invitationAdded']);


const getPath = (str, obj) => str.split('.').reduce((o, i) => o[i], obj);

export default class MockPubSub {
  constructor() {
    getPubSub.default = () => this;
    this.db = {};
  }

  __serialize(dynamicSerializer) {
    const channels = Object.keys(this.db).sort();
    const snapshot = {};
    for (let i = 0; i < channels.length; i++) {
      const channel = channels[i];
      const doc = this.db[channel];
      const channelFields = fieldsToSerialize[channel];
      if (!channelFields) {
        throw new Error(`BAD MOCK: No fieldsToSerialize for pubsub channel ${channel}`);
      }
      snapshot[channel] = dynamicSerializer.toStatic(doc, channelFields, {constant: unsortables.has(channel)});
      // we don't care about the order, so make it repeatable
      const customSorts = fieldsToSort[channel] || [];
      const sortFields = customSorts.concat('channelId');
      snapshot[channel].sort((a, b) => {
        for (let ii = 0; ii < sortFields.length; ii++) {
          const sortField = sortFields[ii];
          const aVal = getPath(sortField, a);
          const bVal = getPath(sortField, b);
          if (aVal === bVal) continue;
          return aVal > bVal;
        }
        return 0;
      });
    }
    return snapshot;
  }

  publish(channel, message) {
    const [channelName, channelId] = channel.split('.');
    this.db[channelName] = this.db[channelName] || [];
    this.db[channelName].push({channelId, message});
  }
}

