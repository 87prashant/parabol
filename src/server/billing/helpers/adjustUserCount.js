import getRethink from 'server/database/rethinkDriver'
import {
  ADD_USER,
  AUTO_PAUSE_USER,
  PAUSE_USER,
  REMOVE_USER,
  UNPAUSE_USER
} from 'server/utils/serverConstants'
import stripe from 'server/billing/stripe'
import shortid from 'shortid'
import {toEpochSeconds} from 'server/utils/epochTime'
import {sendSegmentIdentify} from 'server/utils/sendSegmentEvent'

const changePause = (inactive) => (orgIds, userId) => {
  const r = getRethink()
  return r({
    user: r
      .table('User')
      .get(userId)
      .update({inactive}),
    organizationUser: r
      .table('OrganizationUser')
      .getAll(userId, {index: 'userId'})
      .filter({removedAt: null})
      .update({inactive})
  })
}

const addUser = async (orgIds, userId) => {
  const r = getRethink()
  const {organizations, organizationUsers} = await r({
    organizationUsers: r
      .table('OrganizationUser')
      .getAll(userId, {index: 'userId'})
      .orderBy(r.desc('newUserUntil'))
      .coerceTo('array'),
    organizations: r
      .table('Organization')
      .getAll(r.args(orgIds))
      .coerceTo('array')
  })
  const docs = orgIds.map((orgId) => {
    const oldOrganizationUser = organizationUsers.find(
      (organizationUser) => organizationUser.orgId === orgId
    )
    const organization = organizations.find((organization) => organization.id === orgId)
    return {
      id: shortid.generate(),
      inactive: false,
      joinedAt: new Date(),
      // continue the grace period from before, if any OR set to the end of the invoice OR (if it is a free account) no grace period
      newUserUntil:
        (oldOrganizationUser && oldOrganizationUser.newUserUntil) ||
        organization.periodEnd ||
        new Date(),
      orgId,
      removedAt: null,
      role: null,
      userId
    }
  })
  return r.table('OrganizationUser').insert(docs)
}

const deleteUser = (orgIds, userId) => {
  const r = getRethink()
  return r
    .table('OrganizationUser')
    .getAll(userId, {index: 'userId'})
    .filter((row) => r.expr(orgIds).contains(row('orgId')))
    .update({
      removedAt: new Date()
    })
}

const typeLookup = {
  [ADD_USER]: addUser,
  [AUTO_PAUSE_USER]: changePause(true),
  [PAUSE_USER]: changePause(true),
  [REMOVE_USER]: deleteUser,
  [UNPAUSE_USER]: changePause(false)
}

export default async function adjustUserCount (userId, orgInput, type, options = {}) {
  const r = getRethink()
  const now = new Date()

  const orgIds = Array.isArray(orgInput) ? orgInput : [orgInput]
  const dbAction = typeLookup[type]
  await dbAction(orgIds, userId)
  const orgs = await r
    .table('Organization')
    .getAll(r.args(orgIds), {index: 'id'})
    .merge((organization) => ({
      quantity: r
        .table('OrganizationUser')
        .getAll(organization('id'), {index: 'orgId'})
        .filter({
          inactive: false,
          removedAt: null
        })
        .count()
    }))
  const prorationDate = toEpochSeconds(type === REMOVE_USER ? options.prorationDate : now)
  const hooks = orgs.reduce((arr, org) => {
    const {stripeSubscriptionId} = org
    if (stripeSubscriptionId) {
      arr.push({
        id: shortid.generate(),
        stripeSubscriptionId: org.stripeSubscriptionId,
        prorationDate,
        type,
        userId
      })
    }
    return arr
  }, [])
  // wait here to make sure the webhook finds what it's looking for
  await r.table('InvoiceItemHook').insert(hooks)
  const stripePromises = orgs.reduce((arr, org) => {
    const {quantity, stripeSubscriptionId} = org
    if (stripeSubscriptionId) {
      arr.push(
        stripe.subscriptions.update(stripeSubscriptionId, {
          proration_date: prorationDate,
          quantity
        })
      )
    }
    return arr
  }, [])

  await Promise.all(stripePromises)
  // publish any changes to user traits (like tier counts) to segment:
  await sendSegmentIdentify(userId)
}
