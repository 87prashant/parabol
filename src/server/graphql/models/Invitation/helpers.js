import shortid from 'shortid';
import makeSecretToken from '../../../utils/makeSecretToken';
import ms from 'ms';
import sendEmail from '../../../email/sendEmail';
import r from '../../../database/rethinkDriver';

export const getTeamNameInvitedBy = async (teamId, userId) => {
  const promises = [
    r.table('Team').get(teamId).pluck('name'),
    r.table('UserProfile').get(userId).pluck('preferredName')
  ];

  const [{name: teamName}, {preferredName: invitedBy}] = await Promise.all(promises);
  return {teamName, invitedBy};
};

export const resolveSentEmails = async (sendEmailPromises, invitees, invitations) => {
  const inviteeErrors = [];
  const invitationsToStore = [];
  for (let i = 0; i < sendEmailPromises.length; i++) {
    const promise = sendEmailPromises[i];
    const resolvedPromise = await promise;
    if (resolvedPromise) {
      invitationsToStore.push(invitations[i]);
    } else {
      inviteeErrors.push(invitees[i]);
    }
  }
  return {inviteeErrors, invitationsToStore};
};

export const makeInvitations = (invitees, teamId) => {
  const now = new Date();
  const tokenExpiration = Date.now() + ms('30d');
  return invitees.map(invitee => {
    const {email, task, fullName} = invitee;
    return {
      id: shortid.generate(),
      teamId,
      createdAt: now,
      isAccepted: false,
      inviteToken: makeSecretToken(email, tokenExpiration),
      fullName,
      email,
      task
    };
  });
};

export const sendInvitations = (invitedBy, teamName, invitations) => {
  return invitations.map(invitation => {
    const {email, task, fullName, inviteToken} = invitation;
    const emailData = {
      invitedBy,
      teamName,
      inviteToken,
      email,
      task,
      fullName
    };
    return sendEmail('inviteTeamMember', emailData);
  });
};

