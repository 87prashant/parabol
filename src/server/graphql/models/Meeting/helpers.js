import r from '../../../database/rethinkDriver';

export const getMeetingById = async id => await r.table('Meeting').get(id);

export const getMeetingByTeamId = async teamId =>
 await r.db('actionDevelopment').table('Meeting').filter({ teamId });
