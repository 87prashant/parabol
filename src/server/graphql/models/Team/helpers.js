import r from 'server/database/rethinkDriver';

export const getTeamById = async id => await r.table('Team').get(id);
