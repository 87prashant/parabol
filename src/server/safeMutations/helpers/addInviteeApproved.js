import getRethink from 'server/database/rethinkDriver';
import shortid from 'shortid';
import {INVITEE_APPROVED} from 'universal/utils/constants';

const addInviteeApproved = async (inviteeEmail, inviter) => {
  const r = getRethink();
  const now = new Date();
  const {inviterName, userId, teamId, teamName, orgId} = inviter;
  const notification = {
    id: shortid.generate(),
    type: INVITEE_APPROVED,
    startAt: now,
    orgId,
    userIds: [userId],
    inviteeEmail,
    inviterName,
    teamId,
    teamName
  };
  await r.table('Notification').insert(notification);
  return {[userId]: [notification]};
};

export default addInviteeApproved;
