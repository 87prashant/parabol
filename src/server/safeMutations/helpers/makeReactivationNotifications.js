import tmsSignToken from 'server/utils/tmsSignToken';
import {ADD_TO_TEAM, REJOIN_TEAM} from 'universal/utils/constants';


const makeReactivationNotifications = (reactivatedUsers, teamMembers, inviter) => {
  const {teamId, teamName, inviterName, userId} = inviter;
  const restOfTeamUserIds = teamMembers
    .filter((m) => m.isNotRemoved === true && m.id !== userId)
    .map((m) => m.userId);
  const notificationsToSend = {};
  reactivatedUsers.forEach((user) => {
    const {preferredName, id: reactivatedUserId, tms} = user;

    // make a notification to the person being reactivated
    notificationsToSend[reactivatedUserId] = [{
      authToken: tmsSignToken({sub: reactivatedUserId}, tms),
      inviterName,
      teamId,
      teamName,
      type: ADD_TO_TEAM
    }];

    // make a notification for the other team members annoucing the reactivation
    const rejoinNotification = {
      teamName,
      preferredName,
      type: REJOIN_TEAM
    };
    restOfTeamUserIds.forEach((notificationUserId) => {
      notificationsToSend[notificationUserId] = [rejoinNotification];
    });
  });
  return notificationsToSend;
};

export default makeReactivationNotifications;
