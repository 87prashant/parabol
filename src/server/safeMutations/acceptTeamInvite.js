import adjustUserCount from 'server/billing/helpers/adjustUserCount';
import getRethink from 'server/database/rethinkDriver';
import addUserToTMSUserOrg from 'server/safeMutations/addUserToTMSUserOrg';
import insertNewTeamMember from 'server/safeMutations/insertNewTeamMember';
import getTeamInviteNotifications from 'server/safeQueries/getTeamInviteNotifications';
import {auth0ManagementClient} from 'server/utils/auth0Helpers';
import {getUserId} from 'server/utils/authorization';
import getPubSub from 'server/utils/getPubSub';
import {ADD_USER} from 'server/utils/serverConstants';
import tmsSignToken from 'server/utils/tmsSignToken';
import {
  ADD_TO_TEAM, ADDED, INVITATION_REMOVED, JOIN_TEAM, NEW_AUTH_TOKEN, NOTIFICATIONS_ADDED, NOTIFICATIONS_CLEARED,
  TEAM_ADDED, TEAM_MEMBER
} from 'universal/utils/constants';
import toTeamMemberId from 'universal/utils/relay/toTeamMemberId';

const publishRemovedInvitations = (expiredEmailInvitations, operationId) => {
  expiredEmailInvitations.forEach((invitation) => {
    const {teamId} = invitation;
    const invitationRemoved = {invitation};
    getPubSub().publish(`${INVITATION_REMOVED}.${teamId}`, {invitationRemoved, operationId});
  });
};

const publishClearNotifications = (expireInviteNotificationIds, userId) => {
  if (expireInviteNotificationIds.length > 0) {
    const notificationsCleared = {deletedIds: expireInviteNotificationIds};
    getPubSub().publish(`${NOTIFICATIONS_CLEARED}.${userId}`, {notificationsCleared});
  }
};

const publishJoinTeamNotifications = (teamId, teamName, user) => {
  const notificationsAdded = {
    notifications: [{
      type: JOIN_TEAM,
      teamId,
      teamName,
      preferredName: user.preferredName || user.email
    }]
  };
  getPubSub().publish(`${NOTIFICATIONS_ADDED}.${teamId}`, {notificationsAdded});
};

const acceptTeamInvite = async (teamId, authToken, email, subOptions = {}) => {
  const r = getRethink();
  const now = new Date();
  const {operationId} = subOptions;
  const userId = getUserId(authToken);
  const teamMemberId = toTeamMemberId(teamId, userId);
  const {team: {orgId, name: teamName}, user} = await r({
    team: r.table('Team').get(teamId).pluck('orgId', 'name'),
    user: r.table('User').get(userId)
  });
  const userOrgs = user.userOrgs || [];
  const userTeams = user.tms || [];
  const userInOrg = Boolean(userOrgs.find((org) => org.id === orgId));
  const tms = [...userTeams, teamId];
  const {expiredEmailInvitations, expireInviteNotificationIds, teamMember} = await r({
    // add the team to the user doc
    userUpdate: addUserToTMSUserOrg(userId, teamId, orgId),
    teamMember: insertNewTeamMember(userId, teamId),
    // find all possible emails linked to this person and mark them as accepted
    expiredEmailInvitations: r.table('Invitation')
      .getAll(email, {index: 'email'})
      .update({
        acceptedAt: now,
        // flag the token as expired so they cannot reuse the token
        tokenExpiration: new Date(0),
        updatedAt: now
      }, {returnChanges: true})('changes')('new_val').default([]),
    expireInviteNotificationIds: getTeamInviteNotifications(orgId, teamId, [email])
      .delete({returnChanges: true})('changes')('old_val')('id')
      .default([])
  });

  if (!userInOrg) {
    await adjustUserCount(userId, orgId, ADD_USER);
  }
  auth0ManagementClient.users.updateAppMetadata({id: userId}, {tms});

  publishRemovedInvitations(expiredEmailInvitations, operationId);
  publishClearNotifications(expireInviteNotificationIds, userId);
  publishJoinTeamNotifications(teamId, teamName, user);

  getPubSub().publish(`${TEAM_MEMBER}.${teamId}`, {teamMember: {teamMemberId, type: ADDED}, ...subOptions});

  // Send the new team member a welcome & a new token
  const newAuthToken = tmsSignToken(authToken, tms);
  const addedToTeam = {
    // no ID to rule out permanent notification. they already know, they are the ones who did it!
    startAt: now,
    type: ADD_TO_TEAM,
    teamId
  };
  const teamAdded = {
    notification: addedToTeam,
    teamId,
    teamMemberId
  };
  getPubSub().publish(`${TEAM_ADDED}.${userId}`, {teamAdded, ...subOptions});
  getPubSub().publish(`${NEW_AUTH_TOKEN}.${userId}`, {newAuthToken});

  return {
    authToken: newAuthToken,
    teamId,
    teamMemberId
  };
};

export default acceptTeamInvite;
