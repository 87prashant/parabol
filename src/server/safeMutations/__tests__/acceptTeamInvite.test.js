import DynamicSerializer from 'dynamic-serializer';
import MockDate from 'mockdate';
import MockPubSub from 'server/__mocks__/MockPubSub';
import mockAuthToken from 'server/__tests__/setup/mockAuthToken';
import MockDB from 'server/__tests__/setup/MockDB';
import {__now} from 'server/__tests__/setup/mockTimes';
import fetchAndSerialize from 'server/__tests__/utils/fetchAndSerialize';
import newInvitee from 'server/__tests__/utils/newInvitee';
import * as adjustUserCount from 'server/billing/helpers/adjustUserCount';
import getRethink from 'server/database/rethinkDriver';
import acceptTeamInvite from 'server/safeMutations/acceptTeamInvite';
import {auth0ManagementClient} from 'server/utils/auth0Helpers';
import {ADD_USER} from 'server/utils/serverConstants';
import * as tmsSignToken from 'server/utils/tmsSignToken';
import {TEAM_INVITE} from 'universal/utils/constants';

MockDate.set(__now);
console.error = jest.fn();

describe('acceptTeamInvite', () => {
  beforeEach(() => {
    auth0ManagementClient.users.updateAppMetadata.mockReset();
  });

  test('adds an invitee who was not previously in the org', async () => {
    // SETUP
    adjustUserCount.default = jest.fn();
    tmsSignToken.default = jest.fn(() => 'FAKEENCODEDJWT');
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const mockPubSub = new MockPubSub();
    const invitee = newInvitee('acceptTeamInvite1');
    await mockDB.init()
      .newUser({name: 'inviteeGuy', tms: [], userOrgs: [], email: invitee.email})
      .newNotification(undefined, {type: TEAM_INVITE, email: invitee.email});
    const teamId = mockDB.context.team.id;
    const inviteeUser = mockDB.context.user;
    const authToken = mockAuthToken(inviteeUser);
    const orgId = mockDB.context.organization.id;
    auth0ManagementClient.__initMock(mockDB.db);
    const tms = [teamId];
    // TEST
    await acceptTeamInvite(teamId, authToken, inviteeUser.email);

    // VERIFY
    const db = await fetchAndSerialize({
      user: r.table('User').get(inviteeUser.id),
      teamMember: r.table('TeamMember').get(`${inviteeUser.id}::${teamId}`),
      notification: r.table('Notification').getAll(orgId, {index: 'orgId'}).orderBy('userIds').coerceTo('array'),
      invitation: r.table('Invitation').getAll(inviteeUser.email, {index: 'email'}).orderBy('tokenExpiration').coerceTo('array')
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
    expect(mockPubSub.__serialize(dynamicSerializer)).toMatchSnapshot();
    expect(adjustUserCount.default).toHaveBeenCalledWith(inviteeUser.id, orgId, ADD_USER);
    expect(auth0ManagementClient.users.updateAppMetadata).toHaveBeenCalledWith({id: inviteeUser.id}, {tms});
  });
  test('adds an invitee who was previously in the org', async () => {
    // SETUP
    adjustUserCount.default = jest.fn();
    tmsSignToken.default = jest.fn(() => 'FAKEENCODEDJWT');
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const mockPubSub = new MockPubSub();
    await mockDB.init()
      .newTeam({orgId: mockDB.context.organization.id})
      .newNotification(undefined, {type: TEAM_INVITE, email: mockDB.db.user[1].email});
    const teamId = mockDB.context.team.id;
    const inviteeUser = mockDB.db.user[1];
    const authToken = mockAuthToken(inviteeUser);
    const orgId = mockDB.context.organization.id;
    auth0ManagementClient.__initMock(mockDB.db);
    const tms = [mockDB.db.team[0].id, teamId];

    // TEST
    await acceptTeamInvite(teamId, authToken, inviteeUser.email);

    // VERIFY
    const db = await fetchAndSerialize({
      user: r.table('User').get(inviteeUser.id),
      teamMember: r.table('TeamMember').get(`${inviteeUser.id}::${teamId}`),
      notification: r.table('Notification').getAll(orgId, {index: 'orgId'}).orderBy('userIds').coerceTo('array'),
      invitation: r.table('Invitation').getAll(inviteeUser.email, {index: 'email'}).orderBy('tokenExpiration').coerceTo('array')
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
    expect(mockPubSub.__serialize(dynamicSerializer)).toMatchSnapshot();
    expect(adjustUserCount.default).toHaveBeenCalledTimes(0);
    expect(auth0ManagementClient.users.updateAppMetadata).toHaveBeenCalledWith({id: inviteeUser.id}, {tms});
  });
});
