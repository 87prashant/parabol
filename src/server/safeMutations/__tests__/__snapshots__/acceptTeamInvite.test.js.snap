// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`acceptTeamInvite adds an invitee who was not previously in the org 1`] = `
Object {
  "invitation": Array [],
  "notification": Array [],
  "teamMember": Object {
    "checkInOrder": 4,
    "email": "invitee@mockTeam+acceptTeamInvite1.co",
    "id": 3,
    "isCheckedIn": null,
    "isFacilitator": true,
    "isLead": false,
    "isNotRemoved": true,
    "picture": "https://placeimg.com/100/100/animals",
    "preferredName": "inviteeGuy",
    "teamId": 1,
    "userId": 0,
  },
  "user": Object {
    "cachedAt": 2017-01-01T07:00:00.000Z,
    "createdAt": 2017-01-01T07:00:00.000Z,
    "email": "invitee@mockTeam+acceptTeamInvite1.co",
    "emailVerified": false,
    "id": 0,
    "identities": Array [],
    "inactive": false,
    "lastLogin": 2017-01-01T07:00:00.000Z,
    "lastSeenAt": 2017-01-01T07:00:00.000Z,
    "name": "inviteeGuy",
    "picture": "https://placeimg.com/100/100/animals",
    "preferredName": "inviteeGuy",
    "tms": Array [
      1,
    ],
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userOrgs": Array [
      Object {
        "id": 2,
        "role": null,
      },
    ],
    "welcomeSentAt": 2017-01-01T07:00:00.000Z,
  },
}
`;

exports[`acceptTeamInvite adds an invitee who was not previously in the org 2`] = `
Object {
  "newAuthToken": Array [
    Object {
      "channelId": 0,
      "message": Object {
        "newAuthToken": 5,
      },
    },
  ],
  "notificationsAdded": Array [
    Object {
      "channelId": 0,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "authToken": "FAKEENCODEDJWT",
              "orgId": 2,
              "startAt": 2017-01-01T08:00:00.000Z,
              "teamId": 1,
              "teamName": "Team America",
              "type": "ADD_TO_TEAM",
            },
          ],
        },
      },
    },
    Object {
      "channelId": 1,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "preferredName": "inviteeGuy",
              "teamId": 1,
              "teamName": "Team America",
              "type": "JOIN_TEAM",
            },
          ],
        },
      },
    },
  ],
  "notificationsCleared": Array [
    Object {
      "channelId": 0,
      "message": Object {
        "notificationsCleared": Object {
          "deletedIds": Array [
            6,
          ],
        },
      },
    },
  ],
  "teamMemberAdded": Array [
    Object {
      "channelId": 1,
      "message": Object {
        "teamMemberAdded": Object {
          "teamMember": Object {
            "checkInOrder": 9,
            "email": "invitee@mockTeam+acceptTeamInvite1.co",
            "id": 3,
            "isCheckedIn": null,
            "isFacilitator": true,
            "isLead": false,
            "isNotRemoved": true,
            "picture": "https://placeimg.com/100/100/animals",
            "preferredName": "inviteeGuy",
            "teamId": 1,
            "userId": 0,
          },
        },
      },
    },
  ],
}
`;

exports[`acceptTeamInvite adds an invitee who was previously in the org 1`] = `
Object {
  "invitation": Array [],
  "notification": Array [],
  "teamMember": Object {
    "checkInOrder": 6,
    "email": "vision@avengers.org",
    "id": 5,
    "isCheckedIn": null,
    "isFacilitator": true,
    "isLead": false,
    "isNotRemoved": true,
    "picture": "https://placeimg.com/100/100/animals",
    "preferredName": "Vision",
    "teamId": 2,
    "userId": 0,
  },
  "user": Object {
    "cachedAt": 2017-01-01T07:00:00.000Z,
    "createdAt": 2017-01-01T07:00:00.000Z,
    "email": "vision@avengers.org",
    "emailVerified": false,
    "id": 0,
    "identities": Array [],
    "inactive": false,
    "lastLogin": 2017-01-01T07:00:00.000Z,
    "lastSeenAt": 2017-01-01T07:00:00.000Z,
    "name": "Jonas",
    "picture": "https://placeimg.com/100/100/animals",
    "preferredName": "Vision",
    "tms": Array [
      1,
      2,
    ],
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userOrgs": Array [
      Object {
        "id": 3,
        "role": null,
      },
      Object {
        "id": 4,
        "role": null,
      },
    ],
    "welcomeSentAt": 2017-01-01T07:00:00.000Z,
  },
}
`;

exports[`acceptTeamInvite adds an invitee who was previously in the org 2`] = `
Object {
  "newAuthToken": Array [
    Object {
      "channelId": 0,
      "message": Object {
        "newAuthToken": 7,
      },
    },
  ],
  "notificationsAdded": Array [
    Object {
      "channelId": 0,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "authToken": "FAKEENCODEDJWT",
              "orgId": 3,
              "startAt": 2017-01-01T08:00:00.000Z,
              "teamId": 2,
              "teamName": "Team America",
              "type": "ADD_TO_TEAM",
            },
          ],
        },
      },
    },
    Object {
      "channelId": 2,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "preferredName": "Vision",
              "teamId": 2,
              "teamName": "Team America",
              "type": "JOIN_TEAM",
            },
          ],
        },
      },
    },
  ],
  "notificationsCleared": Array [
    Object {
      "channelId": 0,
      "message": Object {
        "notificationsCleared": Object {
          "deletedIds": Array [
            8,
          ],
        },
      },
    },
  ],
  "teamMemberAdded": Array [
    Object {
      "channelId": 2,
      "message": Object {
        "teamMemberAdded": Object {
          "teamMember": Object {
            "checkInOrder": 1,
            "email": "vision@avengers.org",
            "id": 5,
            "isCheckedIn": null,
            "isFacilitator": true,
            "isLead": false,
            "isNotRemoved": true,
            "picture": "https://placeimg.com/100/100/animals",
            "preferredName": "Vision",
            "teamId": 2,
            "userId": 0,
          },
        },
      },
    },
  ],
}
`;
